L10n

egens opened the pull request:
Another big one here. 99% of text ready. I've added Russian and English-debug l10ns. En-debug adds ### to all text fields - if someone english-speaking wants to help testing. I've added json to https://crowdin.com/project/toolkit-for-ynab I think it's better to have some internal testing before moving this to public.

blargity commented on src/common/res/features/shared/main.js:
```diff
@@ -105,6 +105,30 @@ ynabToolKit.shared = new function() {
 
+    // TODO Maybe add universal function.
+    // Usage: declension(daysNumber, {nom: 'день', gen: 'дня', plu: 'дней'});
+    this.declension = function (locale, num, cases) {
```
Knowing about languages I know what declension is, but it'd be great if we could put a comment above that explains why this exists and what it's for. Arabic is a good example of a language with a shocking amount of odd declension that would be good to reference.

Monolingual English speakers have a tendency to assume that all languages much just be singular or plural like English is. :-)

blargity commented:
I'm seriously impressed @egens! I've read through the code, just going to test now.

blargity commented:
First off, let me say this was so much fun to review. I had always imagined implementing this, but knew I'd probably never get around to it. Thank you so much!

I managed to find some bugs:

- It looks like the change you made in #161 breaks a lot of the localisation when you move around now that I've merged it. I reverted that one commit on my local copy (only) and it started localising a lot more of the strings. I think unfortunately if we want to stick with the new way (I do, personally) then it'll mean going through and updating the observe ```if``` statements in this feature.

- I loaded a budget up on English Debug and got this as Age of Money: 

![screen shot 2016-02-08 at 11 09 48 pm](https://cloud.githubusercontent.com/assets/1531750/12885320/31efd3b4-ceb9-11e5-8bde-2a8428e88184.png)

Not sure why. If I had to guess I'd say it's probably because English Debug has no declension function?

- I managed to get the month name to say Feb instead of Фев while I was in Russian mode. I think it happened on initial load of the budget. I'll keep an eye out and see if I can give you more concrete reproduction steps.

- Many of the buttons we're injecting into the UI aren't localised. Maybe we could figure out a fancy build way to specify what the strings are in each feature, then pull them all together into a combined JSON file similar to what we do with settings. What do you think would be best here?

Thank you again, you rock!

egens commented:
@blargity I've already found some little bugs - mostly text that isn't localized. The logic is monstrous and wants some refactoring, but I can't find a better way for this. I realised one, while implementing this - doing l10n via outer extension for such dynamic app is damn tedious.

I'll look into the bugs you've found, thanks.

An automatic way of features l10n is great, but seems way to hard. For now I'm just doing this
```
-          label.textContent = "Selected Transactions Total";
+          label.textContent = (ynabToolKit.l10nData && ynabToolKit.l10nData.Accounts.Header.Selected) || "Selected Transactions Total";
```

blargity commented:
Yeah, you're probably right that it'd be better to bake them in to the features themselves.

Merge should be relatively simple as each person would be editing a separate part of the english JSON file. Let's just keep doing what you've been doing.

blargity commented on src/common/res/features/l10n/en.json:
```diff
@@ -0,0 +1,423 @@
+          "Cover" : ". Cover the overspending from other categories or you can’t trust your budget balances!",
+          "YouRe" : "You're ",
+          "ShortOfGoal" : " short of your goal for the month. Budget another ",
+          "ToKeepTrack" : " to keep on track."
```
There may not be a good way to do this at all with this library, but the ideal for things like "You're ____ short of your goal for the month. Budget another ______ to keep on track." would be to put a format string in there so if the order of the words needs to change in the translation, the translator can change it.

egens replied:
The thing is there are lines that are splitted by this chunks in element.contents() and my contentSetter.setSeveral works with them nicely, but translation should be splitted too.

shama commented:
Hello! YNAB is using [ember-i18n](https://github.com/jamesarosen/ember-i18n) for i18n and we have most of the templates mapped (we just haven't gotten to the part that switches the language yet).

Here is an export of the current english locale: https://gist.github.com/shama/971c0dc769622c82acba

You might be able to get away with replacing all the current translations with:
```js
Ember.I18n.translations = { /* ... locales ... */ }
```
or individually setting them:
```js
Ember.I18n.translations["app.done"] = "Готово"
```
and then re-render the views.

Hopefully this helps! Cheers! :)

egens commented:
@shama the dumbest feeling I ever had... Where you were all this time?!??! :)

@blargity I'm gonna to make it from scratch...

blargity commented:
@egens, yup, I reckon we should start sourcing translations from the community!

shama commented:
@egens Ha sorry, I just noticed you all working on this perusing your issues this morning. It's amazing how fast you all are developing this project. :D

egens commented:
@shama could you please explain me the thing why there are such things?
```
Ember.I18n.translations['budget.category']
>function(t){return e.replace(/\{\{\{\s*(.*?)\s*\}\}\}/g,function(e,n){return i(t,n)}).replace(/\{\{\s*(.*?)\s*\}\}/g,function(e,n){return d(i(t,n))})}
Ember.I18n.translations['budget.category']()
>"CATEGORY"
```
JSON pretty stringify `JSON.stringify(Ember.I18n.translations, null, 2)` returns 536 items without those functions.

I'm able to get them through this code. It returns 631 items, last of them are 'Missing translation' so there are 615 items.
```
function isFunction(functionToCheck) {
 var getType = {};
 return functionToCheck && getType.toString.call(functionToCheck) === '[object Function]';
}
Object.keys(Ember.I18n.translations).map(function(k){
  var item = Ember.I18n.translations[k];
  try {
    if (isFunction(item)) {    
      return item();
    }
    else {
      return item;
    }
  }
  catch(err) {
    console.log(err);
  }
})
```
That gist contains 619 items. Is it internal thing? Is there an ember function or some way to get it?

shama commented:
@egens Some of the additional items are because I pulled them from the `feature/exports` branch I'm working on which has some i18n additions.

The function ones are because ember-i18n uses simplified handlebar templates (basically a simple regex) to replace variables. `Did you mean {{word}}?` Where `word` gets passed in later: `Ember.I18n.t('app.didYouMean', { word: 'foo' })`. Anywhere you see triple stashes `{{{word}}}` it means the value won't be escaped. We try to use those sparingly and ensure we are escaping any value manually before passing down to the template.

Looping through `Object.keys(Ember.I18n.translations)` like you're doing is the best way to get the current keys.

egens commented:
I reworked old logic, added python script for crowdin integration. Seems to work good.

@shama thanks, I understand. Could you suggest a way to get those template strings from the app itself?

shama commented:
@egens All the templates strings are initially raw. Then as ember-i18n loads them, it will compile it and replace the raw value with the compiled as an optimization.

If you look at the source of `appmain.[hash].js`, the module `./locale/en` (or search for `./en.coffee` or `app.done`) has the latest strings.

During runtime, there might be a way to load that module for the strings but I'm not sure how in production. Or if you can clone `Ember.I18n.translations` before it renders any, you'll have the raw list.

egens commented:
I found that ember is changing it internal translation object while user is moving around so I added reloading of l10n into observer. Looks fine but need testing.

Also some better way of l10n injection should be found or maybe views reloader because now l10n of some thing is applied only after user change that views.

shama commented:
@egens As for reloading the views, see this comment: https://github.com/blargity/toolkit-for-ynab/issues/182#issuecomment-181981146 Once you have access to the route, you might be able to call [refresh](http://emberjs.com/api/classes/Ember.Route.html#method_refresh) on it to reload the route.

We currently don't do this atm nor have an API for it. We just reload the window (as when you change currency settings on your budget). So I'm not 100% sure if `.refresh()` will work but that is what I plan on doing when finishing the i18n implementation down the road.

Thanks for the offer for missing string info. We are aware of the current missing strings. The shared library uses a different i18n library and not all of the strings there are i18nized yet.

egens commented:
@shama thanks, I'll look closer soon.

egens commented:
@blargity I accidentially added api key. Removed it from repo just now. Can you get if from crowdin yourself? L10ns now can be downloaded via `./get_l10ns KEY`

blargity commented:
Yup, I can do that, thanks for the heads up.

blargity commented:
@egens Actually, should we just commit the .json files to the repo so anyone can build it, but you need to get an API key to pull new translations down and update the existing ones?

egens commented:
@blargity I think you're right, json files should be there. Fixing this.

blargity commented:
@egens, just wanted to let you know that I enabled CrowdIn's CNAME support, so you can now use the URL http://translate.toolkitforynab.com.

Invite link is: http://translate.toolkitforynab.com/project/toolkit-for-ynab/invite

The old one will still work too.

egens commented:
@blargity :+1: 
I'm still struggling with not-localized-on-start problem.

blargity commented:
No worries, let me know if there's anything I can do to help.

egens commented:
@blargity done, mb not the best way, but works for me.

blargity commented:
Cool, I'll have a play, thanks!

Just so I know, if we want to introduce new strings (like in the Reports we have Debts, Assets, and Net Worth) could you document where we add them? I'm assuming we'd just add a line to en.json and the next time you update the file they'd go up into CrowdIn?

blargity commented:
This is really awesome @egens! Thank you so much for all your work on this.

There were only a few places I noticed non-localised text (using Czech l10n):

- My Account Link: 

![screen shot 2016-02-14 at 4 44 31 pm](https://cloud.githubusercontent.com/assets/1531750/13032094/424f8238-d33a-11e5-9605-b3ff0ee963aa.png)

- Calendar control:

![screen shot 2016-02-14 at 4 45 25 pm](https://cloud.githubusercontent.com/assets/1531750/13032099/6057f3dc-d33a-11e5-93f3-fac312b40dfe.png)

- Month Names in some Select boxes:

![screen shot 2016-02-14 at 4 45 56 pm](https://cloud.githubusercontent.com/assets/1531750/13032106/7a25ff98-d33a-11e5-8a77-e828dbf743aa.png)

- The displayed month name in the Budget screen:

![screen shot 2016-02-14 at 4 46 57 pm](https://cloud.githubusercontent.com/assets/1531750/13032112/9d45e132-d33a-11e5-84ad-4f5331375a4e.png)

That being said, this is certainly better than no localisation, so I'm happy to merge in and we can keep ironing things like this out later? Or would you rather have a crack now? What do you reckon?

egens commented:
@blargity I've added section to 'how to make features' file about l10n, could you check if all is clear there?

I've improved Reports l10n in terms of adding strings to localize. Went through app's interface and added strings not localized by YNAB to localize it through my old logic. 

These things are not localized by Ember too:

<img width="329" alt="2016-02-14 12 50 42" src="https://cloud.githubusercontent.com/assets/748807/13033774/063fc844-d334-11e5-910d-aa05b36ae84f.png">
<img width="215" alt="2016-02-14 12 45 04" src="https://cloud.githubusercontent.com/assets/748807/13033775/0a32008e-d334-11e5-92ac-00eab74d7b29.png">
<img width="262" alt="2016-02-14 14 30 57" src="https://cloud.githubusercontent.com/assets/748807/13033777/16c8ca8a-d334-11e5-925f-bce538b06a80.png">
<img width="253" alt="2016-02-14 13 07 21" src="https://cloud.githubusercontent.com/assets/748807/13033779/25fb96fe-d334-11e5-9c15-d713318e43a2.png">
<img width="263" alt="2016-02-14 13 07 30" src="https://cloud.githubusercontent.com/assets/748807/13033782/2e79c4f4-d334-11e5-893a-26b10a98bbe2.png">
<img width="480" alt="2016-02-14 15 09 50" src="https://cloud.githubusercontent.com/assets/748807/13033781/2ad9cfa6-d334-11e5-9f46-81e4f1f018ac.png">
<img width="226" alt="2016-02-14 13 09 03" src="https://cloud.githubusercontent.com/assets/748807/13033788/4ded2a92-d334-11e5-9a2d-2c7f803360b5.png">
<img width="288" alt="2016-02-14 12 33 21" src="https://cloud.githubusercontent.com/assets/748807/13033789/4f67b4dc-d334-11e5-90a1-b35e801c8521.png">

egens commented:
Some more untranslated.

There is "app.edit" string, but it doesn't attach
<img width="339" alt="2016-02-14 23 45 05" src="https://cloud.githubusercontent.com/assets/748807/13036325/70c1cad2-d375-11e5-846c-fb53fb7c3fa0.png">

<img width="340" alt="2016-02-14 23 46 02" src="https://cloud.githubusercontent.com/assets/748807/13036329/803e7e38-d375-11e5-8ac9-55d349cf5d27.png">

blargity commented:
Cool! I'm going to go ahead and pull it in as is. We can keep tweaking later, but this will definitely work for people almost 100%.
